### PR TITLE
Check for connection status in finalizer

### PR DESCRIPTION
**Pull request by Yuras**
It is fix for #225 
It also contains related patch to cleanup test suit, please let me know if you want me to send it separately.

### CODE REVIEW

**Comment by Yuras** [comment]
@cdepillabout Could you please check that it fixes the issue for you.

File: http-client/Network/HTTP/Client/Connection.hs [hunk 1]
```diff
@@ -93,7 +93,13 @@ makeConnection r w c = do
     -- already closed connection.
     closedVar <- newIORef False
 
-    _ <- mkWeakIORef istack c
+    let close = do
+          closed <- readIORef closedVar
```
Comment (snoyberg) [inline http-client/Network/HTTP/Client/Connection.hs]: I'm not certain it's necessary, but to avoid a potential race condition, what about using `atomicModifyIORef` to read the closed state and immediately write back `True`?

Comment (snoyberg) [inline http-client/Network/HTTP/Client/Connection.hs]: I realize that's a change to how the code works now.

File: http-client/test-nonet/Network/HTTP/ClientSpec.hs [hunk 2]
```diff
@@ -28,13 +36,13 @@ redirectServer inner = bracket
         (N.runTCPServer (N.serverSettingsTCPSocket lsocket) app)
         (const $ inner port)
   where
-    app ad = do
-        _ <- forkIO $ forever $ N.appRead ad
-        forever $ do
+    app ad = Async.race_
+        (forever (N.appRead ad))
+        (silentIOError $ forever $ do
```
Comment (snoyberg) [inline http-client/test-nonet/Network/HTTP/ClientSpec.hs]: What's the motivation behind this change? It's unclear why it's happening.

File: http-client/Network/HTTP/Client/Connection.hs (hunk 1 above)
Comment (Yuras) [inline http-client/Network/HTTP/Client/Connection.hs]: done

File: http-client/test-nonet/Network/HTTP/ClientSpec.hs (hunk 2 above)
Comment (Yuras) [inline http-client/test-nonet/Network/HTTP/ClientSpec.hs]: It is described briefly in the 9a39b4b87943ec5e7fdffe3292195dd8bcf51caa commit message.

Basically, `forkIO $ forever (N.appRead ad)` thread is never killed, so it performs reading from already closed socket. To trigger it, add `performGC` at the end of test suit to force cached connections to be released -- as a result write loop fails (resource vanished), `app ad` exits, server socket is closed, `appRead ad` fails with either resource vanished or bad descriptor.

In any case, I'm open for better solution.

Comment (snoyberg) [inline http-client/test-nonet/Network/HTTP/ClientSpec.hs]: Nah, that seems fine, thanks for the pointer. Let's wait for Travis.

**Comment by cdepillabout** [comment]
I'm not going to be able to test this until Tuesday.  But the fix looks pretty close to what I described in #225, which did work for me, so I'd bet that this fix is good :+1.

**Comment by snoyberg** [comment]
Travis failure on GHC 7.4.2

**Comment by snoyberg** [comment]
Thanks!
